crates/shared/src/token_info.rs

```diff
@@ -1,173 +1,229 @@
+        );
+
+        Ok(TokenInfo {
+            decimals: classify_error(decimals)?,
+            symbol: classify_error(symbol)?,
+        })
+    }
+}
+
+fn classify_error<T>(result: Result<T, MethodError>) -> Result<Option<T>, Error> {
+    match result {
+        Ok(value) => Ok(Some(value)),
+        Err(err) => match EthcontractErrorType::classify(&err) {
+            EthcontractErrorType::Node => Err(Error(err.to_string())),
+            EthcontractErrorType::Contract => Ok(None),
+        },
+    }
+}
+
 #[async_trait]
 impl TokenInfoFetching for TokenInfoFetcher {
-    async fn get_token_infos(&self, addresses: &[H160]) -> HashMap<H160, TokenInfo> {
-        let mut batch = CallBatch::new(self.web3.transport());
-        let futures = addresses
-            .iter()
-            .map(|address| {
-                let erc20 = ERC20::at(&self.web3, *address);
-                (
-                    erc20.methods().decimals().batch_call(&mut batch),
-                    erc20.methods().symbol().batch_call(&mut batch),
-                )
-            })
-            .collect::<Vec<_>>();
-
-        batch.execute_all(MAX_BATCH_SIZE).await;
-        let mut resolved_futures = Vec::with_capacity(futures.len());
-        for (decimals, symbol) in futures {
-            resolved_futures.push((decimals.await, symbol.await));
+    async fn get_token_info(&self, address: H160) -> Result<TokenInfo, Error> {
+        let info = self.fetch_token(address).await;
+        if let Err(err) = &info {
+            tracing::debug!(?err, token = ?address, "failed to fetch token info");
         }
-        addresses
-            .iter()
-            .zip(resolved_futures)
-            .map(|(address, (decimals, symbol))| {
-                if decimals.is_err() {
-                    tracing::trace!("Failed to fetch token info for token {}", address);
-                }
-                (
-                    *address,
-                    TokenInfo {
-                        decimals: decimals.ok(),
-                        symbol: symbol.ok(),
-                    },
-                )
-            })
-            .collect()
+
+        info
+    }
+
+    async fn get_token_infos(&self, addresses: &[H160]) -> HashMap<H160, TokenInfo> {
+        futures::future::join_all(addresses.iter().copied().map(|address| async move {
+            let info = self.fetch_token(address).await;
+            if let Err(err) = &info {
+                tracing::debug!(?err, token = ?address, "failed to fetch token info");
+            }
+
+            (address, info.unwrap_or_default())
+        }))
+        .await
+        .into_iter()
+        .collect()
     }
 }
 
+type SharedTokenInfo = Shared<BoxFuture<'static, Result<TokenInfo, Error>>>;
+
 pub struct CachedTokenInfoFetcher {
-    inner: Box<dyn TokenInfoFetching>,
-    cache: Arc<Mutex<HashMap<H160, TokenInfo>>>,
+    inner: Arc<dyn TokenInfoFetching>,
+    cache: Arc<Mutex<HashMap<H160, SharedTokenInfo>>>,
```

Comment:
We might have to keep an eye on memory consumption here.
Storing a whole future instead of a `String` and a `u8` for A LOT of tokens might be a problem. Maybe it makes sense to have sth like:
```rust
enum CacheEntry {
    Resolved(TokenInfo),
    InFlight(SharedTokenInfoRequest)
}
```
No need to change in this PR, though.

Reply:
> Storing a whole future instead of a String and a u8 for A LOT of tokens might be a problem

Its a boxed future though, so it should just be a ptr + vtable, which is smaller than `TokenInfo`.

Reply:
Wow, to my surprise, `SharedTokenInfo` is smaller than a `TokenInfo` (without counting the data that exists on the heap): https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4d4fb54c802e8db652c6608eb9c83d35.

This is likely because of a combination of boxing and null-pointer optimization.

With that in mind, I imagine that this type is fairly space-optimized.

Reply:
> Its a boxed future though, so it should just be a ptr + vtable, which is as big as a TokenInfo.

Yeah. The handle to the future is very small but you'll still keep the actual allocation for the future around which I was mainly worried about.

Reply:
It shouldn't though right?

Internally, `Shared` has an `UnsafeCell` to:

```
enum FutureOrOutput<Fut: Future> {
    Future(Fut),
    Output(Fut::Output),
}
```

Here, `Fut = BoxFuture`, so its 16 bytes on x86_64 (ptr + vtable). AFAIU, once the unsafe cell gets set to `FutureOrOutput::Output`, then the future will get dropped right?

Reply:
Ahh, true. I assumed the `SharedFuture` would basically keep the future alive the whole time and simply have a special case for returning a resolved value instead of swapping the future allocation for the resolved result. 👌

Reply:
For example here:

```rust
use futures::FutureExt;
use std::{
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

struct Foo;

impl Future for Foo {
    type Output = i32;
    
    fn poll(self: Pin<&mut Self>, _: &mut Context) -> Poll<Self::Output> {
        Poll::Ready(42)
    }
}

impl Drop for Foo {
    fn drop(&mut self) {
        println!("Foo::drop");
    }
}

fn main() {
    let foo = Foo.boxed().shared();

    {
        println!("cloning handles");
        for i in 0..3 {
            let answer = futures::executor::block_on(foo.clone());
            println!("answer {i}: {answer}");
        }
    }
    
    // We now have a handle to `foo` which was resolved, if we await it, we
    // shouldn't drop the future again.
    println!("foo is resolved");
    let _ = futures::executor::block_on(foo);
}
```

Output:
```
cloning handles
Foo::drop
answer 0: 42
answer 1: 42
answer 2: 42
foo is resolved
```

The future gets dropped the first time it resolves, and only dropped once.

The only issue would be if there are "dangling" futures in the `HashMap`. This can be fixed by removing dangling `Shared` futures that aren't resolved. I also don't suspect this will happen in practice either (especially since follow up token info requests will drive the `Shared` future forward).